### ISSUE TITLE
Open a single request

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#33]
* As an intake analyst
* I want to be able to open request
* so that the request can than be assigned to another member of IAO for action or redirected to another public body

**Assumptions/Scope**	
This story does not include the Contact Applicant function. That will be in a separate story.

Letter creation is out of scope. Document generation and storage are out of scope.
Saving is a separate story

Sending Acknowledgement letter is not included

Does not include AXIS integration.

**Acceptance Criteria**	

**Scenario 1: Click 'Open Request'**
* GIVEN an intake analyst is on an the detailed view of an 'UNOPENED' or 'INTAKE IN PROGRESS' request
https://jacklyn808742.invisionapp.com/console/FOI---Intake-ckpq19frj032r013q6gpm8t1m/ckpx6mtxp0124013ee0rldl83/play
* AND all the mandatory fields have been filled
* AND only 1 ministry is selected
* WHEN the analyst clicks the 'Open Request' button
* AND any changes they made will be saved
* THEN the request status will change to 'Open'
* AND the analyst will be on the request view

**Scenario 2: Generate File Number'**
* GIVEN an intake analyst is on an the detailed view of an 'UNOPENED' or 'INTAKE IN PROGRESS' request
https://jacklyn808742.invisionapp.com/console/FOI---Intake-ckpq19frj032r013q6gpm8t1m/ckpx6mtxp0124013ee0rldl83/play
* AND all the mandatory fields have been filled
* WHEN the analyst clicks the 'Open Request' button
* THEN a file number will be generated with this convention HTH-2021-12345
* AND the file number will be displayed at the top of request

**Scenario 3: Generate Days Remaining'**
* GIVEN an intake analyst is on an the detailed view of an 'UNOPENED' or 'INTAKE IN PROGRESS' request
https://jacklyn808742.invisionapp.com/console/FOI---Intake-ckpq19frj032r013q6gpm8t1m/ckpx6mtxp0124013ee0rldl83/play
* AND all the mandatory fields have been filled
* WHEN the analyst clicks the 'Open Request' button
* THEN the days remaining from the Start Date will be generated
* AND the days remaining will be displayed at the top of request

**Scenario 4: Click 'Open Request' with multiple ministries selected**
* GIVEN an intake analyst is on an the detailed view of an 'UNOPENED' or 'INTAKE IN PROGRESS' request
https://jacklyn808742.invisionapp.com/console/FOI---Intake-ckpq19frj032r013q6gpm8t1m/ckpx6mtxp0124013ee0rldl83/play
* AND all the mandatory fields have been filled
* AND more than 1 ministry is selected
* WHEN the analyst clicks the 'Open Request' button
* THEN a request will become 'Open' for each ministry selected
* AND the analyst will be on the request view for the request based on alphabetical order

**Scenario 4: Click 'Open Request' with multiple ministries selected 2**
* GIVEN an intake analyst has opened multiple requests at once
* WHEN the analyst goes to the Request Queue
* THEN all the requests opened will be listed with a status of "Open"

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.     [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by lmullane** [comment]
@JHarrietha-AOT, please review the scenarios and see if they match with how the designs would work.

I'd also like to suggest the secondary button say 'Close Request' instead of 'Redirect Request.' The button could then be used to 'Close Request' for other use cases (i.e. applicant withdraws request). Thoughts?

@m-prodan, @arielleandrews, question for you - should an analyst be able to close a request, without sending the correspondence to the applicant? I believe we discussed this, and the answer was yes.

Wondering if WHEN an analyst closes a request, that a contact applicant view should automatically open?

cc @mpilchar

**Comment by arielleandrews** [comment]
@lmullane correct, closing a request should not be dependent on sending correspondence to the applicant. We wouldn't close a request without attempting to contact the applicant (probably several times), but there are times we simply aren't able to do this - no email or address provided, phone number out of service, etc.

**Comment by m-prodan** [comment]
@lmullane  agree with Arielle. Intake may ask why we would just have a close button, as there tends to be some nuance as to why we decide to close a request, but we can build that into a secondary page - close request shouldn't automatically close a request, I think there does need to be a secondary screen to explain why it was closed (ie: applicant withdrew, request redirected, does not meet s. 5/no response from applicant, etc..) - this would be important for stat tracking.

As I type all this out, I wonder if closing a request needs to be a separate story?

**Comment by lmullane** [comment]
@JHarrietha-AOT, I've simplified this story to be just opening a single request. Note, the toast so the analyst gets a success message when the email is sent and the request is opened. Let's discuss at refinement if toast is the common pattern for success messages.

**Comment by lmullane** [comment]
Dependency - need to use AXIS file number convention as later this will need to integrate.

**Comment by sumathi-thirumani-aot** [comment]
@JHarrietha-AOT  @m-prodan @arielleandrews @lmullane 
Below scenarios would lead in race condition,
1. There is a possibility of more than one intake analyst review and assign request.
2. More than one intake analyst can work on request; as assignment is mandated only during submit of request.

**Comment by abin-aot** [comment]
@lmullane @m-prodan  - Between the planning, I was trying to understand the scenario that while we go to Prod/MVP release with only 1 pilot ministry. Then while processing the FOI submission if multiple

**Comment by m-prodan** [comment]
Thanks @abin-aot - assume you are meaning that until the pilot Ministry is selected and ready to go - for purposes of a test, the application will just default to the first Ministry selected alphabetically when multiple are picked?

**Comment by abin-aot** [comment]
@m-prodan - No, Scenario 4 will not be there. We will open for all selected ministries. If we restrict it to a single one, then we will not be able to open the FOI submission that has more than 1. Technically, nothing is stopping us to implement this.

**Comment by m-prodan** [comment]
Ok we should discuss quickly at standup tomorrow.

**Comment by lmullane** [comment]
@abin-aot, AC updated as discussed.

**Comment by m-prodan** [comment]
@divyav-aot @abin-aot @sumathi-thirumani-aot 

Will need to send this one back for review - found issues quickly:

First, when I submitted request on online form, the received date on the queue had tomorrow's date:

![image.png](https://images.zenhubusercontent.com/6081a4c82f0c363dddb9c1a0/e8ea7ea9-7278-4751-89d7-b1b1ec52c071)

Now as you can see in review request view, the actual received date correctly shows here:
![image.png](https://images.zenhubusercontent.com/6081a4c82f0c363dddb9c1a0/06100893-4c84-4e43-b258-1c41218b7a35)

But additional problems once I 'Opened' the request - the received and start date both changed to August 16th (2 days ago)
![image.png](https://images.zenhubusercontent.com/6081a4c82f0c363dddb9c1a0/4fae5142-6579-4c4d-ac62-90048545aca7)

And then it doesn't look like the days remaining is calculated properly:
![image.png](https://images.zenhubusercontent.com/6081a4c82f0c363dddb9c1a0/9307ff0a-cfb8-4aa1-9c18-1ff52e4ea4cb)

And lastly - received date on the queue now shows as yesterday - August 17th (likely how days remaining is being calculated)
![image.png](https://images.zenhubusercontent.com/6081a4c82f0c363dddb9c1a0/f8e2c5b9-4009-4e25-be2a-9aeb180286f1)

So I would say this needs to go back for fixing, as acceptance criteria 3 is not met, days remaining not calculating properly. Let me know if I need to submit additional tickets for the other bugs noted. I will re-iterate at standup tomorrow.

**Comment by m-prodan** [comment]
Wrapping up QA and closing this ticket - nice work on the fixes! The issue I talked about earlier at stand up, I am unable to replicate - I don't even know how I did it. I'll write up a bug ticket and point to the specific request this occurred in for perhaps some analysis, or something to watch for in the future in case we are able to produce it again. We should also update the acronyms for the request file number, to be more consistent with AXIS for now, but I think this can be handled as a separate issue.
